core/src/mast/debuginfo/mod.rs

```diff
@@ -1,7 +1,258 @@
+    pub fn decorator(&self, decorator_id: DecoratorId) -> Option<&Decorator> {
+        self.decorators.get(decorator_id)
+    }
+
+    /// Returns the before-enter decorators for the given node.
+    pub fn before_enter_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        self.node_decorator_storage.get_before_decorators(node_id)
+    }
+
+    /// Returns the after-exit decorators for the given node.
+    pub fn after_exit_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        self.node_decorator_storage.get_after_decorators(node_id)
+    }
+
+    /// Returns decorators for a specific operation within a node.
+    pub fn decorators_for_operation(
+        &self,
+        node_id: MastNodeId,
+        local_op_idx: usize,
+    ) -> &[DecoratorId] {
+        self.op_decorator_storage
+            .decorator_ids_for_operation(node_id, local_op_idx)
+            .unwrap_or(&[])
+    }
+
+    /// Returns decorator links for a node, including operation indices.
+    pub fn decorator_links_for_node(
+        &self,
+        node_id: MastNodeId,
+    ) -> Result<DecoratedLinks<'_>, DecoratorIndexError> {
+        self.op_decorator_storage.decorator_links_for_node(node_id)
+    }
+
+    // DECORATOR MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds a decorator and returns its ID.
+    pub fn add_decorator(&mut self, decorator: Decorator) -> Result<DecoratorId, MastForestError> {
+        self.decorators.push(decorator).map_err(|_| MastForestError::TooManyDecorators)
+    }
+
+    /// Adds node-level decorators (before_enter and after_exit) for the given node.
+    pub fn add_node_decorators(
+        &mut self,
+        node_id: MastNodeId,
+        before_enter: &[DecoratorId],
+        after_exit: &[DecoratorId],
+    ) {
+        self.node_decorator_storage
+            .add_node_decorators(node_id, before_enter, after_exit);
+    }
+
+    /// Adds decorator info for operations within a node.
+    pub fn add_decorator_info_for_node(
+        &mut self,
+        node_id: MastNodeId,
+        decorators_info: Vec<(usize, DecoratorId)>,
+    ) -> Result<(), crate::mast::debuginfo::decorator_storage::DecoratorIndexError> {
+        self.op_decorator_storage.add_decorator_info_for_node(node_id, decorators_info)
+    }
+
+    /// Returns the decorator with the given ID, if it exists (mutable).
+    pub fn decorator_mut(&mut self, decorator_id: DecoratorId) -> Option<&mut Decorator> {
+        if decorator_id.to_usize() < self.decorators.len() {
+            Some(&mut self.decorators[decorator_id])
+        } else {
+            None
+        }
+    }
+
+    /// Clears all decorator information while preserving error codes.
+    /// This is used when rebuilding decorator information from nodes.
+    pub fn clear_decorators(&mut self) {
+        self.op_decorator_storage = OpToDecoratorIds::new();
+        self.node_decorator_storage.clear();
+    }
```

Comment:
As mentioned in one of the above comments: do we not want to clear `self.decorators` here as well?

Reply:
We do not, as explained [here](https://github.com/0xMiden/miden-vm/pull/2366#discussion_r2570079739)